### ISSUE TITLE
Benchmark/Expected time for a N-digits address?

### ISSUE THREAD

**Issue by stevefan1999-personal** [cathugger/mkp224o#27]
I want to know the average case of getting the desired vanity address on various CPUs, compilers, OS, etc.

I compiled my mkp224o with clang 7 and march=native, in Debian Buster and with 2x Xeon E5-2670

hello:
```bash
./mkp224o hello -B -S 5 -j 16 -d keys
>calc/sec:5095952.833017, succ/sec:0.000000, rest/sec:159.888078, elapsed:0.100070sec
hello23twa7k536qggxfeqm4orwohwlca3ln6f43b4splcym57msaxid.onion
```

hellow:
```bash
./mkp224o hellow -B -S 5 -j 16 -d keys
...
>calc/sec:7007205.907981, succ/sec:0.000000, rest/sec:0.000000, elapsed:15.111171sec                           [27/9839]
hellowthsxpoy4pq4vensv2drz6qv562d3sfqg5pk4nwqnik4hii53ad.onion
```

hellowo:
```bash
./mkp224o hellowo -B -S 5 -j 16 -d keys
...
>calc/sec:7318805.372164, succ/sec:0.000000, rest/sec:0.000000, elapsed:155.117082sec
hellowoy3ahurx54upkhdut2ww6kwsmzrrtve6imdul4f5yr4kpsb2id.onion
```

Should I need more samples so that we can calculate the variance? This is pretty random at best.

**Comment by cathugger** [comment]
Nature of the way it works is, in fact, random.
Single char is 5 bits (base32) so probability of hitting right thing shrinks 2^5=32 times by each added character.
You'd need to run it longer to calculate averange.
I'd recommend first benchmarking differences with lower char count as higher hit rate will demand less total benchmark time to smooth out noise.
Best way to benchmark for purpose of speed improvement (like picking different compilers/processing backends) is looking at calc/sec stat, longer char counts shouldn't affect it.
Properly averanged over longer period of time hit benchmark shouldn't diverge significantly from theorethical prediction of 32 times less hit rate each added character.
Due to probablistic nature of it, there can always be surprises and disappointments, even if you know averange time it can find it sooner or take longer.

**Comment by YoRyan** [comment]
The math (a geometric discrete probability distribution) tells us that, on average, it will take `32^(# of chars)` iterations to find a desired address. So take that number and divide by your calcs/sec to get the expected run time until your first match.

For example, my i7-3770 does ~15,400,000 calcs/sec. To find a 7-character prefix, the anticipated run time is
```
32^7/15400000 = 2231.15 secs = 37 mins
```

And indeed, it took me 40 minutes to find my first address.

**Comment by stevefan1999-personal** [comment]
> 
> 
> The math (a geometric discrete probability distribution) tells us that, on average, it will take `32^(# of chars)` iterations to find a desired address. So take that number and divide by your calcs/sec to get the expected run time until your first match.
> 
> For example, my i7-3770 does ~15,400,000 calcs/sec. To find a 7-character prefix, the anticipated run time is
> 
> ```
> 32^7/15400000 = 2231.15 secs = 37 mins
> ```
> 
> And indeed, it took me 40 minutes to find my first address.
> 
> (Note that this compares very favorably with [shallot](https://github.com/katmagic/Shallot#performance), which quotes "1 day" for a 1.5GHz CPU and a 7-character prefix.)

Did this means it takes for any x in length of digits, it should theoretically take 2^(5x) tries to have at least one match? This is huge 🤔

**Comment by YoRyan** [comment]
Correct! But it's just an average. You could get especially lucky or unlucky and find something sooner or later.

**Comment by cathugger** [comment]
>(Note that this compares very favorably with shallot, which quotes "1 day" for a 1.5GHz CPU and a 7-character prefix.)

"1.5GHz CPU" isn't very descriptive, though, and that table is quite outdated ([history](https://github.com/katmagic/Shallot/commits/master/README.asciidoc)), so I'd say it's just difference of CPU powers, and docs being outdated.
One of the reasons I didn't put any stats in README.txt in the first place.

>Did this means it takes for any x in length of digits, it should theoretically take 2^(5x) tries to have at least one match?

No. It means, that **on average** you should get one match every 2^(5x)=32^x tries. That's a little bit different. You can get very lucky and hit much sooner. Or very unlucky and hit much later. I've seen both happening.
Statistically, yeah, you'll get one match every 32^x tries (if using only one filter of x chars), but for that to show you'll need to average over much more than one hit, so I'm not sure if such expectation formula is very trustworthy when N=1.
This stuff is sorta confusing, but I just don't wanna overpromise such things in readme, especially when most users want only one or few addresses.

**Comment by scribblemaniac** [comment]
I agree with the math that's been presented here so far. If you want to be a bit more conservative, you can calculate the number of trials *t* it would take to get at least 1 match with *n* characters and a probability *p* with the formula `t = log(1-p)/log(1-1/32^n)`. This comes from the distribution `X~Binomial(t, 1/32^n)` and `P(X>=1)=p`. Here's a table of what the number of trials looks like for some values common values. The number of digits are on the left and the probability of finding at least one success along the top:

|  | 50% | 80% | 90% | 95% | 99% |
|---|---|---|---|---|---|
| **2** | 710 | 1648 | 2357 | 3067 | 4714 |
| **3** | 22713 | 52738 | 75450  | 98163 | 150900 |
| **4** | 726818 | 1687618 | 2414435 | 3141252 | 4828869 |
| **5** | 23258160 | 54003775 | 77261934 | 100520094 | 154523868 |
| **6** | 744261118 | 1728120799 | 2472381917 | 3216643035 | 4944763834 |
| **7** | 23816355775 | 55299865590 | 79116221365 | 102932577139 | 158232442729 |

The way to read this is like so: "It is expected that you will find at least one match for a 2 character search after 710 trials 50% of the time." So while yes on average it will take about ~24 billion trials to find a match for a 7 digit prefix, it will take ~158 billion trials to be 99% sure that you will find a match.

**Comment by cathugger** [comment]
This is very cool, I knew something like this was doable but didn't remember exact maths for this.
I almost wanna write small utility to calculate these and print into terminal now.
That'd probably be pretty useful, to get such table with expected times, from input of calc/sec stat.

**Comment by YoRyan** [comment]
> "1.5GHz CPU" isn't very descriptive, though, and that table is quite outdated ([history](https://github.com/katmagic/Shallot/commits/master/README.asciidoc)), so I'd say it's just difference of CPU powers, and docs being outdated.

It also reflects the fact that mkp224o is a lot faster, especially with the batching code enabled. ;)

> This stuff is sorta confusing, but I just don't wanna overpromise such things in readme, especially when most users want only one or few addresses.

Right, there is always some uncertainty involved, but as we've seen, there is a way to quantify the exact probabilities. It is possible to be more precise than "some time within the next 30 seconds or next 30 years."

**Comment by DadiBit** [comment]
> You can calculate the number of trials _t_ it would take to get at least 1 match with _n_ characters and a probability _p_ with the formula `t = log(1-p)/log(1-1/32^n)`.

In C++ (I think in C too) it's a bit faster `log(1-p)/log(1-1/32)*pow(32,n-1);` rather then `log(1-p)/log(1-1/pow(32,n));`. This is of course a one time calculation before running the program and requires way less the a ms, so it shouldn't matter too much.

**Comment by GhostDog98** [comment]
Based off @scribblemaniac's provided math, I made an easier to use online calculator for estimating it where you can simply paste in your hashrate and it will spit out time estimates. It's all client side javascript based. https://ghostoverflow.gay/tor_calc
Let me know if you want any features added or if you spot any errors.

**Comment by GhostDog98** [comment]
Interestingly the math means the formula scribblemaniac gave runs into an issue after 11 characters, where it jumps to infinity. 
See this desmos graph: https://www.desmos.com/calculator/ds5hnqrsya
<img width="450" height="881" alt="Image" src="https://github.com/user-attachments/assets/1d87a40e-625e-4d89-a6df-8dac1b30d52f" />

**Comment by scribblemaniac** [comment]
> In C++ (I think in C too) it's a bit faster log(1-p)/log(1-1/32)*pow(32,n-1); rather then log(1-p)/log(1-1/pow(32,n));

Those functions are not equivalent.

---

@GhostDog98 Thanks for the tool. It doesn't work with target lengths > 10 characters because Javascript doubles don't have sufficient precision to represent `1-1/32^n` as anything other than 1 with values of n > 10. With `log(1) = 0`, then `x/log(1)` should be undefined. But in the twisted world of Javascript `x/0 = Infinity`. I recommend using `t = log(1-p)/(-1/32^n)` as an approximation for high n values. This comes from the first-order approximation of `log(1-x)`.

A nice additional feature would be to include an numerical input for the number of prefixes being searched for. If k represents the number of prefixes being searched for, then the distribution is `X~Binomial(t, k/32^n)` and the equation is `t = log(1-p)/log(1-k/32^n)` or `t = log(1-p)/(-k/32^n)` for the approximation when n is large.

**Comment by GhostDog98** [comment]
Awesome, cheers for the tips, just got around to uploading that fix and it now has a box for # of patterns searching. For anyone wondering, this is the javascript code for the function described by scribble:
```js
function calculate_trials(number_of_digits, probability_factor) {
        // t = log(1-p)/(-k/32^n)
        let numerator = Math.log(1 - probability_factor);
        let denominator = ((-number_of_search_patterns) / Math.pow(32, number_of_digits));
        return Math.round(numerator / denominator);;
    }
```
Also the table kinda breaks at some higher values, but if you're considering `19,663,525,671,163,648,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000` attempts, I think you're probably okay with that lol. I might fix it at some stage but rn busy with work.
